repository/copier.go

```diff
@@ -113,126 +197,154 @@ func (c *HDFSCopier) freeClient(err error) (wrapped error) {
 	return
 }
 
-// CopyFromRemote copies the file from HDFS to the provided billy Filesystem. If the file exists locally is overridden.
-// If a writer is actually overriding the file on HDFS, we will able to read it, but a previous version of it.
-func (c *HDFSCopier) CopyFromRemote(src, dst string, localFs billy.Filesystem) (err error) {
+func (fs *hdfsFs) initializeClient() (err error) {
+	if fs.client != nil {
+		return
+	}
+
+	fs.client, err = hdfs.New(fs.url)
+	return
+}
+
+func (fs *hdfsFs) Open(path string) (r io.ReadCloser, err error) {
 	defer func() {
-		err = c.freeClient(err)
+		err = fs.freeClient(err)
 	}()
 
-	if err := c.initializeClient(); err != nil {
-		return err
-	}
-
-	bSrc := addBucketName(src, c.bucketSize)
-	rf, err := c.client.Open(path.Join(c.base, bSrc))
-	if os.IsNotExist(err) {
-		return nil
-	}
-	if err != nil {
-		return err
+	if err = fs.initializeClient(); err != nil {
+		return
 	}
-	defer checkClose(rf, &err)
 
-	lf, err := localFs.OpenFile(dst, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, os.FileMode(0644))
+	r, err = fs.client.Open(path)
 	if err != nil {
-		return err
+		return
 	}
-	defer checkClose(lf, &err)
 
-	_, err = io.Copy(lf, rf)
 	return
 }
 
-func (c *HDFSCopier) deleteIfExists(file string) error {
-	if _, err := c.client.Stat(file); os.IsNotExist(err) {
-		return nil
-	} else if err != nil {
-		return err
-	}
-
-	return c.client.Remove(file)
-}
-
-// CopyToRemote copies from the provided billy Filesystem to HDFS. If the file exists on HDFS it will be overridden.
-// If other writer is actually copying the same file to HDFS this method will throw an error because the WORM principle
-// (Write Once Read Many).
-func (c *HDFSCopier) CopyToRemote(src, dst string, localFs billy.Filesystem) (err error) {
+func (fs *hdfsFs) WriteTo(path string) (w io.WriteCloser, err error) {
 	defer func() {
-		err = c.freeClient(err)
+		err = fs.freeClient(err)
 	}()
 
-	bDst := addBucketName(dst, c.bucketSize)
-	p := path.Join(c.base, bDst)
-	if err := c.initializeClient(); err != nil {
-		return err
+	if err = fs.initializeClient(); err != nil {
+		return
 	}
 
-	pCopy := path.Join(c.tmpBase, dst+".copy")
+	if err = fs.client.MkdirAll(filepath.Dir(path), os.FileMode(0644)); err != nil {
+		return
+	}
 
-	lf, err := localFs.Open(src)
-	if os.IsNotExist(err) {
-		return nil
+	if err = fs.DeleteIfExists(path); err != nil {
```

Comment:
If ignore an `error` here maybe we can just call `Remove(path)`, instead calling `fs.Stat` (which is quite expensive) inside `fs.DeleteIfExists`  to check if file exists.

Reply:
Then we have to actually ignore the error. I think it's ok to call `fs.Stat`, given that it's not going to be a bottleneck here (copy is way more expensive and clone is even more expensive) and get to keep the error check.